### PR TITLE
Custom skills sample project

### PR DESCRIPTION
**Pull request by eshaverma**
Project uses 
1) skills-kit source and test code that would soon be internalised into box-node-sdk
2) main index file for lambda
3) helpful error-parse and metadata-card-adaptor classes that users may modify
3) serverless configuration for seamless cloud deployment
4) eslint configuration for formatting
5) async-await syntax support

pending- more documentation & README content

### CODE REVIEW

File: box-node-custom-skills-sample/index.js [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+        const { read: { access_token: boxFileReadToken }, write: { access_token: boxFileWriteToken } } = token;
+        const { fileReadClient, skillInvocations } = new SkillsKit(
+            boxRequestId,
+            skill.id,
+            CUSTOM_SKILL_NAME,
+            source.id,
+            boxFileReadToken,
+            boxFileWriteToken
+        );
+        skillInvocationsClient = skillComponents.skillInvocations;
```
Comment (KarthikShanmugasundaram) [inline box-node-custom-skills-sample/index.js]: skillComponents is not defined.

File: box-node-custom-skills-sample/index.js [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+ * This is the main function that the Lambda will call when invoked.
+ *
+ * @param {Object} event - data from the event, including the payload of the webhook, that triggered this function call
+ * @param {Object} context - additional context information from the request (unused in this example)
+ * @param {Function} callback - the function to call back to once finished
+ * @return {null} - lambda callback
+ */
+module.exports.handler = async (event: Object, context: Object, callback: Function): void => {
+    // log the box skills event recieved from box event pump, to see what data is availabe to you regarding a file on which your skill is invoked.
+    Logger.logDebug(`Box Event received: ${JSON.stringify(event)}`);
```
Comment (KarthikShanmugasundaram) [inline box-node-custom-skills-sample/index.js]: Logger is not defined.

File: box-node-custom-skills-sample/index.js [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+        // @TODO use standard File Metadata Card templates create for skills, to show either faces(timeline), topics, or transcript cards, with your extracted data.
+        const cards = await sampleResultProcessingCode(skillInvocationsClient, source.id);
+        console.debug('cards created as: ', JSON.stringify(cards));
+
+        // save the above result cards to Files Metadata cards, to show on the file preview UI sidebar.
+        await skillInvocationsClient.saveMetadata(cards, SkillsStatusEnum.SUCCESS.value, null, error =>
+            console.error(JSON.stringify(error))
+        );
+    } catch (error) {
+        Logger.logError(`Runtime exception caught: ${error.message}`);
```
Comment (KarthikShanmugasundaram) [inline box-node-custom-skills-sample/index.js]: Logger is not defined.

File: box-node-custom-skills-sample/index.js [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+            source.id,
+            boxFileReadToken,
+            boxFileWriteToken
+        );
+        skillInvocationsClient = skillComponents.skillInvocations;
+
+        // write back an intitial File Metadata card that event processing has started, to show on the file preview UI sidebar.
+        await skillInvocationsClient.savePending();
+
+        // @TODO use the file infromation from the incoming skills request to download the file from box, in one of two ways-
```
Comment (KarthikShanmugasundaram) [inline box-node-custom-skills-sample/index.js]: As this is a sample, May we write the code and just giving the File id as configurable.

File: box-node-custom-skills-sample/metadata-card-adaptor.js [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+        const topicData = [{ text: `Hello World from file ${fileId}` }];
+        const sampleKeywordCard = skillInvocationsClient.TopicCardTemplate(keywordData);
+
+        const transcriptData = [{ text: `Hello World from file ${fileId}`, appears: [{ start: 0, end: 1 }] }, 1];
+        const sampleTranscriptCard = skillInvocationsClient.TranscriptCardTemplate(transcriptData);
+
+        // timeline data can contain an image_url property with a string value of a thumbnail image.
+        // e.g. of a Base 64 encoded URI for an image url:
+        // data:image/jpeg;base64,/9j/4AAQSkZJRgABAQEASABIAAD/2wBDAAcFBQYFBAcGBQYIBwcIChELCgkJChUPEAwRGBU.....
+        sampleThumbnailUrl = 'https://pbs.twimg.com/profile_images/885529357904510976/tM0vLiYS_400x400.jpg';
```
Comment (KarthikShanmugasundaram) [inline box-node-custom-skills-sample/metadata-card-adaptor.js]: Please change it to Box logo.

File: box-node-custom-skills-sample/skills-kit.js [hunk 6]
```diff
@@ -0,0 +1,433 @@
...
+const processCardData = (cardData, duration) => {
+    if (!cardData.text) throw new TypeError(`Missing required 'text' field in ${JSON.stringify(cardData)}`);
+    cardData.type = typeof cardData.image_url === 'string' ? 'image' : 'text';
+    if (duration && !(Array.isArray(cardData.appears) && cardData.appears.length > 0)) {
+        console.warn(
+            `Missing optional 'appears' field in ${JSON.stringify(cardData)} which is list of 'start' and 'end' fields`
+        );
+    }
+};
+/**
```
Comment (KarthikShanmugasundaram) [inline box-node-custom-skills-sample/skills-kit.js]: Please give a line break.

File: box-node-custom-skills-sample/test/skills-kit.test.js [hunk 7]
```diff
@@ -0,0 +1,220 @@
...
+            entries: [{ appears: [{ end: 1.0, start: 0.0 }], text: 'text1', type: 'text' }],
+            invocation: { id: 'mockBoxId', type: 'skill_invocation' },
+            skill: { id: 'mockSkillName', type: 'service' },
+            skill_card_title: { code: 'skills_faces', message: 'Faces' },
+            skill_card_type: 'timeline',
+            status: {},
+            type: 'skill_card'
+        });
+    });
+    test('CardTemplatesInvalidInputDataFormat', () => {
```
Comment (KarthikShanmugasundaram) [inline box-node-custom-skills-sample/test/skills-kit.test.js]: line break

File: box-node-custom-skills-sample/test/skills-kit.test.js [hunk 8]
```diff
@@ -0,0 +1,220 @@
...
+            entries: [{ appears: [{ end: 1.0, start: 0.0 }], text: 'text1', type: 'text' }],
+            invocation: { id: 'mockBoxId', type: 'skill_invocation' },
+            skill: { id: 'mockSkillName', type: 'service' },
+            skill_card_title: { code: 'skills_transcript', message: 'Transcript' },
+            skill_card_type: 'transcript',
+            status: {},
+            type: 'skill_card'
+        });
+    });
+    test('TimelineCardTemplate', () => {
```
Comment (KarthikShanmugasundaram) [inline box-node-custom-skills-sample/test/skills-kit.test.js]: lined break

File: box-node-custom-skills-sample/test/skills-kit.test.js [hunk 9]
```diff
@@ -0,0 +1,220 @@
...
+            entries: [{ appears: [{ end: 1.0, start: 0.0 }], text: 'text1', type: 'text' }],
+            invocation: { id: 'mockBoxId', type: 'skill_invocation' },
+            skill: { id: 'mockSkillName', type: 'service' },
+            skill_card_title: { code: 'skills_my_topics', message: 'my topics' }, // custom title case, default title.message would be 'Topics'
+            skill_card_type: 'keyword',
+            status: {},
+            type: 'skill_card'
+        });
+    });
+    test('TranscriptCardTemplate', () => {
```
Comment (KarthikShanmugasundaram) [inline box-node-custom-skills-sample/test/skills-kit.test.js]: line break

File: box-node-custom-skills-sample/test/skills-kit.test.js [hunk 10]
```diff
@@ -0,0 +1,220 @@
...
+            expect(JSON.parse(err.response.request.body)).toEqual({
+                status: 'success', // default status
+                file: { type: 'file', id: 'mockfileId' },
+                metadata: { cards: [] }, // cards of various types as shown being created in above tests could go here
+                usage: { unit: 'files', value: 1 } // defualt usage
+            });
+            done();
+        });
+    });
+    test('saveMetadataWithDefaultStatusAndNonDefaultUsage', done => {
```
Comment (KarthikShanmugasundaram) [inline box-node-custom-skills-sample/test/skills-kit.test.js]: line break

File: box-node-custom-skills-sample/test/skills-kit.test.js [hunk 11]
```diff
@@ -0,0 +1,220 @@
...
+            expect(JSON.parse(err.response.request.body)).toEqual({
+                status: 'success', // default status
+                file: { type: 'file', id: 'mockfileId' },
+                metadata: { cards: [] }, // cards of various types as shown being created in above tests could go here
+                usage: { unit: 'seconds', value: 10 } // default usage
+            });
+            done();
+        });
+    });
+    test('saveMetadataWithStringUsage', done => {
```
Comment (KarthikShanmugasundaram) [inline box-node-custom-skills-sample/test/skills-kit.test.js]: line break

File: box-node-custom-skills-sample/test/skills-kit.test.js [hunk 12]
```diff
@@ -0,0 +1,220 @@
...
+        try {
+            skillInvocations.saveMetadata([], null, 'badUsage');
+        } catch (err) {
+            expect(typeof err === typeof TypeError);
+            done();
+        }
+    }); // empty cards
+    test('saveMetadataWithDecimalValueUsage', done => {
```
Comment (KarthikShanmugasundaram) [inline box-node-custom-skills-sample/test/skills-kit.test.js]: break

File: box-node-custom-skills-sample/test/skills-kit.test.js [hunk 13]
```diff
@@ -0,0 +1,220 @@
...
+        try {
+            skillInvocations.saveMetadata([], null, { unit: 'seconds', value: 1.7 });
+        } catch (err) {
+            expect(typeof err === typeof TypeError);
+            done();
+        } // non-integer value
+    });
+    test('saveMetadataWithInvalidUnitUsage', done => {
```
Comment (KarthikShanmugasundaram) [inline box-node-custom-skills-sample/test/skills-kit.test.js]: line break

File: box-node-custom-skills-sample/test/skills-kit.test.js [hunk 14]
```diff
@@ -0,0 +1,220 @@
...
+        try {
+            skillInvocations.saveMetadata([], null, { unit: 'badUnit', value: 1 });
+        } catch (err) {
+            expect(typeof err === typeof TypeError);
+            done();
+        } // empty cards
+    });
+    test('saveMetadataWithMissingValueUsage', done => {
```
Comment (KarthikShanmugasundaram) [inline box-node-custom-skills-sample/test/skills-kit.test.js]: line break

File: box-node-custom-skills-sample/test/skills-kit.test.js [hunk 15]
```diff
@@ -0,0 +1,220 @@
...
+    });
+});
+
+describe('fileUtils', () => {
+    const { skillInvocations } = new SkillsKit(boxRequestId, skillId, skillName, fileId, 'readToken', 'writeToken');
+    test('getFileExtensionWithDot', () => {
+        const ext = skillInvocations.getFileExtension('file.ext');
+        expect(ext).toEqual('.ext');
+    });
+    test('getFileExtensionWithoutDot', () => {
```
Comment (KarthikShanmugasundaram) [inline box-node-custom-skills-sample/test/skills-kit.test.js]: line break

File: box-node-custom-skills-sample/test/skills-kit.test.js [hunk 16]
```diff
@@ -0,0 +1,220 @@
...
+        const ext = skillInvocations.getFileExtension('file.ext', true);
+        expect(ext).toEqual('ext');
+    });
+    test('getFileExtensionFromFullFilePath', () => {
```
Comment (KarthikShanmugasundaram) [inline box-node-custom-skills-sample/test/skills-kit.test.js]: line break

File: box-node-custom-skills-sample/test/skills-kit.test.js [hunk 17]
```diff
@@ -0,0 +1,220 @@
...
+        const ext = skillInvocations.getFileExtension('/path/to/file.ext', false);
+        expect(ext).toEqual('.ext');
+    });
+    test('validateFileTypeCapital', () => {
```
Comment (KarthikShanmugasundaram) [inline box-node-custom-skills-sample/test/skills-kit.test.js]: line break

File: box-node-custom-skills-sample/test/skills-kit.test.js [hunk 18]
```diff
@@ -0,0 +1,220 @@
...
+        const valid = skillInvocations.validateFileType('file.EXT', ['.ext']);
+        expect(valid).toEqual(true);
+    });
+    test('validateFileTypeInvalidFileType', () => {
```
Comment (KarthikShanmugasundaram) [inline box-node-custom-skills-sample/test/skills-kit.test.js]: line break

File: box-node-custom-skills-sample/index.js (hunk 1 above)
Comment (eshaverma) [reply box-node-custom-skills-sample/index.js]: thanks! done

File: box-node-custom-skills-sample/index.js (hunk 2 above)
Comment (eshaverma) [reply box-node-custom-skills-sample/index.js]: thanks! done

File: box-node-custom-skills-sample/index.js (hunk 3 above)
Comment (eshaverma) [reply box-node-custom-skills-sample/index.js]: done

File: box-node-custom-skills-sample/index.js (hunk 4 above)
Comment (eshaverma) [reply box-node-custom-skills-sample/index.js]: i changed @TODO to <PLEASE WRITE YOUR OWN CODE HERE> to clearly indicate the Skill developer needs to write their own code here

**Review by KarthikShanmugasundaram — APPROVED** [review]

**Pull request merged**